### ISSUE TITLE
Access Violation Error

### ISSUE THREAD

**Issue by Bciceksoy** [ohrules/CoolerMasterHappyBirthday#1]
Hello, I am trying to test your code but it gives me the same error all the time.

Exception thrown at 0x00007FF9F4509E9F (SDKDLL.dll) in CM_Test.exe: 0xC0000005: Access violation reading location 0x00007FF7602D5F50.

What is the problem? Can you help me? Thanks in advance.

**Comment by osama-h-rana** [comment]
Hi, can you please list in detail how exactly are you trying to make the project?

**Comment by Bciceksoy** [comment]
Sorry for lack of info, I am working in VS 2017 but installed VS 2010 either. Tried on both but 2017 seems gonna work because, the if (IsDevicePlug()) part is working. I added the SDKDLL.h  and it compiles and builds just fine. When it comes to the color changing code SetLedColor(i, 16, 255, 255, 255);

it gives the exception at the live which H(); function appears in main function;

Exception thrown at 0x00007FF9E7379EA3 (SDKDLL.dll) in CM_Test.exe: 0xC0000005: Access violation reading location 0x0000000000000000. occurred

**Comment by osama-h-rana** [comment]
Hmm, I tested this code on a white backlight only keyboard, not sure if this exact code will work on RGB as well. Lemme see if I can find a solution.

Also, may I ask how did you find this repo? Reddit?

**Comment by Bciceksoy** [comment]
Okay huge thanks for the help, yes I found it in this reddit page;
https://www.reddit.com/r/coolermaster/comments/6cnac9/cooler_master_custom_lighting_sdk_example_source/

**Comment by osama-h-rana** [comment]
Did you change line 164 `SetControlDevice(DEV_MKeys_M_White);`  to your specific model? Refer to page 10 of [this document](https://makerhub.coolermaster.com/custom-lighting/assets/sdk/sdk-document.pdf)

**Comment by Bciceksoy** [comment]
Yes, I changed it, it wasn't giving me any info about if (IsDevicePlug()) first then i changed it to DEV_MKeys_L and it gives me info about it now.

**Comment by osama-h-rana** [comment]
My guess is that I missed something that is not needed for white leds but is needed for RGB. I'll follow-up on this tomorrow after reading the docs

**Comment by Bciceksoy** [comment]
Okay thank you, I will be looking forward :)

**Comment by osama-h-rana** [comment]
Have you tried running the code in debug mode (starting from line 165) and stepping over each line to see where exactly the code fails?

**Comment by Bciceksoy** [comment]
It fails at the SetLedColor function. First it failed at line 171 which h() function appears. So, I copied the H function to the main function and it failed at SetLedColor. The problem is SetLedColor function i am sure. Because it goes through this 
```int main()
{
	SetControlDevice(DEV_MKeys_L);
	if (IsDevicePlug())
	{
		printf("YES");
		EnableLedControl(true); //take control of the keyboard
		
		EnableLedControl(false); //give control back
	}
	else
		printf("NO");
	return 0;
}
```

and prints out "YES" when I plug out the keyboard it prints "NO" just fine.

**Comment by osama-h-rana** [comment]
Try this code and tell me the output on the terminal

```
#include <Windows.h>
#include "SDKDLL.h"
#include <iostream>

using namespace std;

int main()
{
	SetControlDevice(DEV_MKeys_M_White);
	if (IsDevicePlug())
	{
		cout << EnableLedControl(true); //take control of the keyboard

		cout << SetFullLedColor(0, 0, 0); //disable all lights

		cout << SetLedColor(0, 0, 127, 127, 127); //set just the escape key to half brightness
		
		cout << EnableLedControl(false); //give control back
	}
	return 0;
}
```

**Comment by Bciceksoy** [comment]
` 'CM_Test.exe' (Win32): Loaded 'C:\Users\Burak\Desktop\CM_Framework-master\CM_Framework-master\x64\Debug\CM_Test.exe'. Symbols loaded.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\ntdll.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\kernel32.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\KernelBase.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Users\Burak\Desktop\CM_Framework-master\CM_Framework-master\x64\Debug\SDKDLL.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\user32.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\win32u.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\gdi32.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\gdi32full.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\msvcp_win.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\ucrtbase.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\advapi32.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\msvcrt.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\sechost.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\rpcrt4.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\ole32.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\combase.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\bcryptprimitives.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\oleaut32.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\setupapi.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\cfgmgr32.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\shlwapi.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\msvcp140d.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\vcruntime140d.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\ucrtbased.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\winspool.drv'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\kernel.appcore.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\pdh.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\hid.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\propsys.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\SHCore.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\IPHLPAPI.DLL'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\bcrypt.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\imm32.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\uxtheme.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\msctf.dll'. Cannot find or open the PDB file.
The thread 0x5290 has exited with code 0 (0x0).
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\devobj.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\wintrust.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\msasn1.dll'. Cannot find or open the PDB file.
'CM_Test.exe' (Win32): Loaded 'C:\Windows\System32\crypt32.dll'. Cannot find or open the PDB file.
Exception thrown at 0x00007FF9F3449F35 (SDKDLL.dll) in CM_Test.exe: 0xC0000005: Access violation reading location 0x00007FF859B016B0.

The program '[900] CM_Test.exe' has exited with code 0 (0x0).
`

By the way, My keyboard just got all black, and then at the `cout << SetLedColor(0, 0, 127, 127, 127);` part it gave the exeption again. I commented out all lines starting with cout exept `cout << EnableLedControl(false);` to get back the previous effect.

**Comment by osama-h-rana** [comment]
Did you make the change to x64 in the project settings?

**Comment by Bciceksoy** [comment]
Okay, It finally worked! And btw, yes i was always in x64.

I wanted to start with clean project. I added your code (the last one) and added the SDKDLL.h header to the project. It wanted the SDKDLL.DLL and SDKDLL64.lib files and i copied them from the original SDK folder. And started the program nothing happened first because there was no sleep!, It happened so fast that i did not see it. I added a 3000 sleep just before the EnableLedControl(false); and it just WORKED !!! 👯‍♂️  ESC key led semi and turned off after 3 seconds and my normal layout came back. Then I just deleted the cout's and it worked again!

I really don't know what happened and what not. Anyway I am so thrilled and HUGE thanks for your help even if I don't know what was happening in the other solution. I guess its murphy or mess solution folder :)

Thank you again, I will be sharing my works...

**Comment by osama-h-rana** [comment]
Awesome! So you think your directory structure was the issue?

**Comment by Bciceksoy** [comment]
I will look into that and share here below 👍  :)

**Comment by Bciceksoy** [comment]
Actually, now i figured out that I had opened that project in VS2010 and was compiling in 2017 and I gathered the .h .dll and .lib files from different examples so it got mixed. And my code was in another example so it just got messy i guess. I tryed about 5-6 codes in the same solution and it just got complicated more. Starting with new solution is really important that's what i learned today :)

Thank you again and sorry for the trouble.
